File: src/ObjectManager/Object.Tes/FilePacks/EsmFile+Fields.cs
using OA.Core;
using System;
using System.Runtime.InteropServices;
using UnityEngine;

namespace OA.Tes.FilePacks
{
    public interface IHaveEDID
    {
        STRVField EDID { get; }
    }

    public interface IHaveMODL
    {
        MODLGroup MODL { get; }
    }

    public class MODLGroup
    {
        public string Value;
        public float Bound;
        public byte[] Textures; // Texture Files Hashes
        public override string ToString() => $"{Value}";

        public MODLGroup(UnityBinaryReader r, int dataSize)
        {
            Value = r.ReadASCIIString((int)dataSize, ASCIIFormat.PossiblyNullTerminated);
        }

        public void MODBField(UnityBinaryReader r, int dataSize)
        {
            Bound = r.ReadLESingle();
        }

        public void MODTField(UnityBinaryReader r, int dataSize)
        {
            Textures = r.ReadBytes(dataSize);
        }

    }

    public struct FormId32<TRecord> where TRecord : Record
    {
        public override string ToString() => $"{Type}:{Id}";
        public readonly uint Id;
        public string Type => typeof(TRecord).Name.Substring(0, 4);
    }

    public struct FormId<TRecord> where TRecord : Record
    {
        public override string ToString() => $"{Type}:{Name}{Id}";
        public readonly uint Id;
        public readonly string Name;
        public string Type => typeof(TRecord).Name.Substring(0, 4);

        public FormId(uint id) { Id = id; Name = null; }
        public FormId(string name) { Id = 0; Name = name; }
        FormId(uint id, string name) { Id = id; Name = name; }
        public FormId<TRecord> AddName(string name) => new FormId<TRecord>(Id, name);
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct ColorRef3 { public byte Red; public byte Green; public byte Blue; public override string ToString() => $"{Red}:{Green}:{Blue}"; }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct ColorRef4 { public byte Red; public byte Green; public byte Blue; public byte Null; public override string ToString() => $"{Red}:{Green}:{Blue}"; public Color32 ToColor32() => new Color32(Red, Green, Blue, 255); }

    public static class ReaderExtension
    {
        public static INTVField ReadINTV(this UnityBinaryReader r, int length)
        {
            switch (length)
            {
                case 1: return new INTVField { Value = r.ReadByte() };
                case 2: return new INTVField { Value = r.ReadLEInt16() };
                case 4: return new INTVField { Value = r.ReadLEInt32() };
                case 8: return new INTVField { Value = r.ReadLEInt64() };
                default: throw new NotImplementedException($"Tried to read an INTV subrecord with an unsupported size ({length})");
            }
        }

        public static DATVField ReadDATV(this UnityBinaryReader r, int length, char type)
        {
            switch (type)
            {
                case 'b': return new DATVField { B = r.ReadLEInt32() != 0 };
                case 'i': return new DATVField { I = r.ReadLEInt32() };
                case 'f': return new DATVField { F = r.ReadLESingle() };
                case 's': return new DATVField { S = r.ReadASCIIString(length, ASCIIFormat.PossiblyNullTerminated) };
                default: throw new InvalidOperationException($"{type}");
            }
        }

        public static STRVField ReadSTRV(this UnityBinaryReader r, int length, ASCIIFormat format = ASCIIFormat.PossiblyNullTerminated)
        {
            return new STRVField { Value = r.ReadASCIIString(length, format) };
        }

        public static FILEField ReadFILE(this UnityBinaryReader r, int length, ASCIIFormat format = ASCIIFormat.PossiblyNullTerminated)
        {
            return new FILEField { Value = r.ReadASCIIString(length, format) };
        }

        public static BYTVField ReadBYTV(this UnityBinaryReader r, int length)
        {
            return new BYTVField { Value = r.ReadBytes(length) };
        }

        public static UNKNField ReadUNKN(this UnityBinaryReader r, int length)
        {
            return new UNKNField { Value = r.ReadBytes(length) };
        }
    }

    public struct STRVField { public string Value; public override string ToString() => Value; }
    public struct FILEField { public string Value; public override string ToString() => Value; }
    public struct INTVField { public long Value; public override string ToString() => $"{Value}"; public UI16Field ToUI16Field() => new UI16Field { Value = (ushort)Value }; }
    public struct DATVField { public bool B; public int I; public float F; public string S; public override string ToString() => "DATV"; }
    public struct FLTVField { public float Value; public override string ToString() => $"{Value}"; }
    public struct BYTEField { public byte Value; public override string ToString() => $"{Value}"; }
    public struct IN16Field { public short Value; public override string ToString() => $"{Value}"; }
    public struct UI16Field { public ushort Value; public override string ToString() => $"{Value}"; }
    public struct IN32Field { public int Value; public override string ToString() => $"{Value}"; }
    public struct UI32Field { public uint Value; public override string ToString() => $"{Value}"; }

    public struct FMIDField<TRecord> where TRecord : Record
    {
        public override string ToString() => $"{Value}";
        public FormId<TRecord> Value;

        public FMIDField(UnityBinaryReader r, int dataSize)
        {
            Value = dataSize == 4 ?
                new FormId<TRecord>(r.ReadLEUInt32()) :
                new FormId<TRecord>(r.ReadASCIIString(dataSize, ASCIIFormat.ZeroPadded));
        }

        public void AddName(string name)
        {
            Value = Value.AddName(name);
        }
    }

    public struct FMID2Field<TRecord> where TRecord : Record
    {
        public override string ToString() => $"{Value1}x{Value2}";
        public FormId<TRecord> Value1;
        public FormId<TRecord> Value2;

        public FMID2Field(UnityBinaryReader r, int dataSize)
        {
            Value1 = new FormId<TRecord>(r.ReadLEUInt32());
            Value2 = new FormId<TRecord>(r.ReadLEUInt32());
        }
    }

    public struct CREFField { public ColorRef4 Color; }

    public struct CNTOField
    {
        public override string ToString() => $"{Item}";
        public uint ItemCount; // Number of the item
        public FormId<Record> Item; // The ID of the item

        public CNTOField(UnityBinaryReader r, int dataSize, GameFormatId format)
        {
            if (format == GameFormatId.TES3)
            {
                ItemCount = r.ReadLEUInt32();
                Item = new FormId<Record>(r.ReadASCIIString(32, ASCIIFormat.ZeroPadded));
                return;
            }
            Item = new FormId<Record>(r.ReadLEUInt32());
            ItemCount = r.ReadLEUInt32();
        }
    }

    public struct BYTVField { public byte[] Value; public override string ToString() => $"BYTS"; }
    public struct UNKNField { public byte[] Value; public override string ToString() => $"UNKN"; }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/EsmFile+Fields.swift
//
//  EsmFile+Fields.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

import Foundation

public protocol IHaveEDID {
    var EDID: STRVField { get }
}

public protocol IHaveMODL {
    var MODL: MODLGroup? { get }
}

public class MODLGroup: CustomStringConvertible {
    public var description: String { return "\(value)" }
    public let value: String
    public var bound: Float? = nil
    public var textures: Data? = nil // Texture Files Hashes

    init(_ r: BinaryReader, _ dataSize: Int) {
        value = r.readASCIIString(dataSize, format: .possibleNullTerminated)
    }

    func MODBField(_ r: BinaryReader, _ dataSize: Int) {
        bound = r.readLESingle()
    }

    func MODTField(_ r: BinaryReader, _ dataSize: Int) {
        textures = r.readBytes(dataSize)
    }
}

public struct FormId32<TRecord> {
    public let id: UInt32
}

public struct FormId<TRecord>: CustomStringConvertible {
    public var description: String { return "\(type):\(name ?? "none")\(id ?? 0)" }
    public let id: UInt32?
    public let name: String?
    public var type: String { let r = "\(TRecord.self)"; return String(r[r.startIndex..<r.index(r.startIndex, offsetBy: 4)])  }
    
    init(_ id: UInt32) { self.id = id ; name = nil }
    init(_ name: String) { id = 0 ; self.name = name }
    init(_ id: UInt32, _ name: String) { self.id = id; self.name = name }
    func adding(name: String) -> FormId<TRecord> { return FormId<TRecord>(id!, name) }
}

public let ColorRef4_empty = ColorRef4(red: 0, green: 0, blue: 0, null: 0)
public func ColorRef4_toColor32(v: ColorRef4) -> CGColor { return CGColor(red: CGFloat(v.red), green: CGFloat(v.green), blue: CGFloat(v.blue), alpha: 255) }
public typealias ColorRef4 = (red: UInt8, green: UInt8, blue: UInt8, null: UInt8)
public typealias ColorRef3 = (red: UInt8, green: UInt8, blue: UInt8)


public let STRVField_empty = STRVField("")
public typealias STRVField = (String)
public typealias FILEField = (String)
public typealias INTVField = (Int64)
public typealias DATVField = (b: Bool?, i: Int32?, f: Float?, s: String?)
public typealias FLTVField = (Float)
public typealias BYTEField = (UInt8)
public typealias IN16Field = (Int16)
public typealias UI16Field = (UInt16)
public typealias IN32Field = (Int32)
public typealias UI32Field = (UInt32)

public struct FMIDField<TRecord>: CustomStringConvertible {
    public var description: String { return "\(value)" }
    public var value: FormId<TRecord>

    init(_ r: BinaryReader, _ dataSize: Int) {
        value = dataSize == 4 ?
            FormId<TRecord>(r.readLEUInt32()) :
            FormId<TRecord>(r.readASCIIString(dataSize, format: .zeroPadded))
    }

    mutating func add(name: String) {
        value = value.adding(name: name)
    }
}

public struct FMID2Field<TRecord>: CustomStringConvertible {
    public var description: String { return "\(value1)x\(value2)" }
    public let value1: FormId<TRecord>
    public let value2: FormId<TRecord>

    init(_ r: BinaryReader, _ dataSize: Int) {
        value1 = FormId<TRecord>(r.readLEUInt32())
        value2 = FormId<TRecord>(r.readLEUInt32())
    }
}

public typealias CREFField = (ColorRef4)

public struct CNTOField: CustomStringConvertible {
    public var description: String { return "\(item)" }
    public let itemCount: UInt32 // Number of the item
    public let item: FormId<Record> // The ID of the item

    init(_ r: BinaryReader, _ dataSize: Int, for format: GameFormatId)
    {
        if format == .TES3 {
            itemCount = r.readLEUInt32()
            item = FormId<Record>(r.readASCIIString(32, format: .zeroPadded))
            return;
        }
        item = FormId<Record>(r.readLEUInt32())
        itemCount = r.readLEUInt32()
    }
}

public typealias BYTVField = (Data)
public typealias UNKNField = (Data)
